src/bedrock/monitor/monitoring_utils.cpp

```diff
@@ -0,0 +1,515 @@
+                                  unsigned& rid) {
+  string target_address = prepare_metadata_request(key, global_memory_hash_ring, local_memory_hash_ring, addr_request_map, mt, rid, "GET");
+
+  if (!target_address.empty()) {
+    prepare_get_tuple(addr_request_map[target_address], key);
+  }
+}
+
+void prepare_metadata_put_request(string& key,
+                                  string& value,
+                                  GlobalHashRing& global_memory_hash_ring,
+                                  LocalHashRing& local_memory_hash_ring,
+                                  unordered_map<string, communication::Request>& addr_request_map,
+                                  MonitoringThread& mt,
+                                  unsigned& rid) {
+  string target_address = prepare_metadata_request(key, global_memory_hash_ring, local_memory_hash_ring, addr_request_map, mt, rid, "PUT");
+
+  if (!target_address.empty()) {
+    prepare_put_tuple(addr_request_map[target_address], key, value, 0);
+  }
+}
+
+void collect_internal_stats(unordered_map<unsigned, GlobalHashRing>& global_hash_ring_map,
+                            unordered_map<unsigned, LocalHashRing>& local_hash_ring_map,
+                            SocketCache& pushers,
+                            MonitoringThread& mt,
+                            zmq::socket_t& response_puller,
+                            shared_ptr<spdlog::logger> logger,
+                            unsigned& rid,
+                            unordered_map<string, unordered_map<string, unsigned>>& key_access_frequency,
+                            unordered_map<string, unordered_map<unsigned, unsigned long long>>& memory_tier_storage,
+                            unordered_map<string, unordered_map<unsigned, unsigned long long>>& ebs_tier_storage,
+                            unordered_map<string, unordered_map<unsigned, pair<double, unsigned>>>& memory_tier_occupancy,
+                            unordered_map<string, unordered_map<unsigned, pair<double, unsigned>>>& ebs_tier_occupancy,
+                            unordered_map<string, unordered_map<unsigned, unsigned>>& memory_tier_access,
+                            unordered_map<string, unordered_map<unsigned, unsigned>>& ebs_tier_access,
+                            unordered_map<unsigned, TierData>& tier_data_map) {
+  unordered_map<string, communication::Request> addr_request_map;
+
+  for (auto it = global_hash_ring_map.begin(); it != global_hash_ring_map.end(); it++) {
+    unsigned tier_id = it->first;
+    auto hash_ring = &(it->second);
+    unordered_set<string> observed_ip;
+
+    for (auto iter = hash_ring->begin(); iter != hash_ring->end(); iter++) {
+      if (observed_ip.find(iter->second.get_ip()) == observed_ip.end()) {
+        for (unsigned i = 0; i < tier_data_map[tier_id].thread_number_; i++) {
+          string key = string(METADATA_IDENTIFIER) + "_" + iter->second.get_ip() + "_" + to_string(i) + "_" + to_string(tier_id) + "_stat";
```

Comment:
Can we have helper methods, perhaps in `metadata.hpp` that automatically create and parse these strings? It seems very brittle to construct and parse them by hand. A mistaken change to this format would probably fail silently and cause this to go haywire.

Reply:
I actually found some bugs about it and fixed them. But I think it makes more sense to make this a separate PR because we also need to fix the server and the routing side's metadata parsing.

Reply:
Agreed about the separate PR!